### PR TITLE
[Core] Ray logs API (1/n)

### PR DESCRIPTION
**Pull request by jon-chuang**
## Why are these changes needed?
Expose HTTP endpoint to retrieve logs from ray cluster

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [inline dashboard/modules/log/log_head.py:184]: TODO: better timeout/failure handling here? (all or nothing or partial result?)

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [inline dashboard/modules/log/log_head.py:112]: Get rid of this

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [inline dashboard/modules/log/log_agent.py:101]: Get rid of this

**Comment by rkooo567** [comment]
Is this ready for review?

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [inline dashboard/modules/log/log_head.py:191]: change to `/api/experimental/`

**Comment by rkooo567** [comment]
Please update the PR description !

File: python/ray/internal/internal_api.py
Comment (rkooo567) [inline python/ray/internal/internal_api.py:320]: Can you move it to experimental folder and create log.py or sth?

File: python/ray/scripts/scripts.py
Comment (rkooo567) [inline python/ray/scripts/scripts.py:1857]: Move it to log.py (after addressing the above coment)

File: python/ray/scripts/scripts.py
Comment (rkooo567) [inline python/ray/scripts/scripts.py:1980]: Two API qusetions.

1. What's happening if we just call ray logs? Does it return the head node list or all log files? (Personally prefer the former).
2. If there's only one match, does it print the log or the index?

**Review by rkooo567 — COMMENTED** [review]
I added the initial comments (but it is not fully carefully reviewed yet).

Also, do you mind splitting PRs to 2~3? 870 lines of code is pretty big!

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:183]: Well, at startup, we should spin until we are connected to the agents

Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:183]: Else it will appear that the logs endpoint doesn't work. Usually it happens in less than 1s. And it is async so we aren't blocking.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:136]: it doesn't use self parameter

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:143]: I think both are prone to error. There isn't a solution if the filename can change arbitrarily. If it breaks in a bad way, we can fix it. Alternatively, we can just return a list of files without categorizing by component.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:138]: I think 10s is more reasonable considering that the SLA is probably more like a few seconds.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:142]: We are either checking each condition for each item in the list, or checking the whole list for each condition. The complexity is the same.

The current way is more readable imo.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:187]: I think if no node id is given, the user should expect the logs index of the whole cluster.

File: python/ray/scripts/scripts.py
Comment (jon-chuang) [reply python/ray/scripts/scripts.py:1980]: I feel that its too verbose and confusing to document the behaviour in the CLI help. Users should discover this behaviour, it feels intuitive IMO.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:50]: I think I will return an empty dir in this case

File: dashboard/modules/log/tests/test_log.py
Comment (jon-chuang) [reply dashboard/modules/log/tests/test_log.py:158]: added test for index / list to catch filename changes.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:143]: added tests to catch filename-related errors by checking returned filenames

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:49]: Hmm this should always exist. Did you see any edge case it doesn't exist?

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:83]: Does it block the event loop or await?

Comment (rkooo567) [reply dashboard/modules/log/log_agent.py:83]: If it blocks the event loop, consider adding asyncio.sleep(0) at the end of the while loop

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:149]:
```suggestion
        def contains_all_filters(log_file_name):
```

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [reply dashboard/modules/log/log_head.py:203]: Also please update the type annotaiton!

File: dashboard/modules/log/tests/test_log.py
Comment (rkooo567) [inline dashboard/modules/log/tests/test_log.py:165]: The test doesn't cover following scenarios

- filtering of list logs
- Actual streaming logs
- Streaming logs when files are actively updated
- Just regular log print

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:181]: Add `runtime_env` too! https://github.com/ray-project/ray/blob/7eb3543e93e7cf3c173c22681f3e3b62ee80f383/python/ray/_private/log_monitor.py#L350

File: dashboard/modules/node/node_consts.py
Comment (rkooo567) [inline dashboard/modules/node/node_consts.py:2]: Can we not change this value? This might affect other parts of dashboard

**Review by rkooo567 — COMMENTED** [review]
Please update the PR description!

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:64]: default is actually to give the whole file.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:49]: It should always exist, but you mentioned to check it just to avoid the uncaught error

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:213]: I think it is useful for SDK and dashboard consumers.

For CLI, I paginate to 10 nodes.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:79]: https://stackoverflow.com/questions/59195598/is-there-an-optimal-message-size-when-streaming-over-grpc#:~:text=It%20depends%20a%20lot%20on,of%20configurations%20including%20mobile%20etc.
https://docs.microsoft.com/en-us/aspnet/core/grpc/performance?view=aspnetcore-6.0

Maybe we could set to 64KB...? Don't think it matters too much...

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py]: I changed to `api/experimental/logs/list`, hope that's fine...

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [reply dashboard/modules/log/log_agent.py:79]: Hmm yeah let's do 64KB. Also no strong opinion here. I was just curious.

File: dashboard/modules/log/tests/test_log.py
Comment (jon-chuang) [reply dashboard/modules/log/tests/test_log.py]: added test for streaming (w/ live update), tailing, actor id, and logs list filtering

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:142]:
```suggestion
        on the given node.
```

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:218]: Is getting the log contents added in the next PR? It seems like we only have a list API

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:219]:
```suggestion
    async def list_logs(self, req):
```
I think the helper function could be `_list_logs`? I am fine with handle_list_logs, so it is up to you

File: dashboard/modules/log/tests/test_log.py
Comment (rkooo567) [inline dashboard/modules/log/tests/test_log.py:188]: What is this break for?

File: src/ray/protobuf/reporter.proto
Comment (rkooo567) [inline src/ray/protobuf/reporter.proto:74]: Can you add comments on each attribute?

File: src/ray/protobuf/reporter.proto
Comment (rkooo567) [inline src/ray/protobuf/reporter.proto:82]: Same here

File: src/ray/protobuf/reporter.proto
Comment (rkooo567) [inline src/ray/protobuf/reporter.proto:93]:
```suggestion
  // Get the list of logs from the cluster.
```

File: src/ray/protobuf/reporter.proto
Comment (rkooo567) [inline src/ray/protobuf/reporter.proto:88]: here too!

**Review by rkooo567 — COMMENTED** [review]
I think it is almost there! Btw, it seems like the current PR doesn't have API for reading a log file although the code is there in agent? Is this intentional?

**Comment by jon-chuang** [comment]
> I think it is almost there! Btw, it seems like the current PR doesn't have API for reading a log file although the code is there in agent? Is this intentional?

Yep! That will be the next PR!

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:57]: Something bad. The gRPC client/stub hangs or will timeout.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:69]: No, because our application code will return upon receiving this metadata without waiting for the reply body.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:218]: yep

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:219]: yep I think _XXX will be for all internal methods. I think handle is a good way to indicate HTTP handler.

File: dashboard/modules/log/tests/test_log.py
Comment (jon-chuang) [reply dashboard/modules/log/tests/test_log.py:188]: The test runs in a loop until the entire block can run without an exception or until timeout.

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:236]: Is this just stylistic?

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:57]: I think that's why its a good idea to have timeout.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:88]: Added a test which will be included in next PR (e2e with streaming through head node)

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [reply dashboard/modules/log/log_agent.py:66]: In this case, should we try/catch all body of the gRPC handler? Relying on timeout seems like a bad idea

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [reply dashboard/modules/log/log_head.py:240]: It is the Python convention I believe (one new line between methods)

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [reply dashboard/modules/log/log_head.py:223]: The thing is we don't follow this anywhere. Imo, doing only partially is worse than not doing it. But it is minor, so I will leave it up to you!

**Comment by rkooo567** [comment]
Is it ready for the final review?

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:240]: sry, my bad, thought you meant to add two lines

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:66]: Actually, my bad, it throws an exception in the client, I handle this in logs_head.py now.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py]: Seems like it is.

**Comment by jon-chuang** [comment]
Should be ready once tests pass

File: dashboard/modules/log/tests/test_log.py
Comment (rkooo567) [reply dashboard/modules/log/tests/test_log.py:188]: Hmm I am confused a little bit... How is the next line executed then?

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [reply dashboard/modules/log/log_head.py:206]: In this case, can you always put the head node node id first and add a test?

**Review by rkooo567 — APPROVED** [review]
Please address the last comments!

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:244]:
```suggestion
            logger.exception(e)
```

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:53]: what's this for?

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:64]:
```suggestion
                f"Could not find log dir at path: {self._dashboard_agent.log_dir}. It is unexpected. Please report an issue to Ray Github."
```

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:51]: can you make these log debug level?

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:54]: debug log

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:75]: debug log

File: dashboard/modules/log/log_agent.py
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:78]: debug log

File: dashboard/modules/log/tests/test_log.py
Comment (jon-chuang) [reply dashboard/modules/log/tests/test_log.py:188]: It doesn't, that's the point. If all the asserts pass, it will break the loop and not retry

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:53]: I guess it is just debugging

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:206]: Wait, I don't understand. So isn't the conclusion that we should display the full index if no node id is specified? Since that is probably what SDK and dashboard want. 

If we want to get head node's, it should be easy to get via its node id.

Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:206]: Ok, so we should have head node's list be sorted as the first entry in the json...

File: dashboard/modules/log/tests/test_log.py
Comment (jon-chuang) [reply dashboard/modules/log/tests/test_log.py:188]: maybe I should remove the loop

File: dashboard/modules/log/log_head.py
Comment (jon-chuang) [reply dashboard/modules/log/log_head.py:206]: It's quite hard to reorder json manually. Also I think we are not guaranteed that consumer will get the dict in order when they do json.loads. Hence, I think we should leave it.

File: dashboard/modules/log/log_agent.py
Comment (jon-chuang) [reply dashboard/modules/log/log_agent.py:56]: From discussion offline:
I think this kills two birds with one stone, making the gRPC handlers observable in the logs.
Testing the behaviour is only a small part of the usefulness.

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Rerunning mac tests now

**Comment by rkooo567** [comment]
cc @jon-chuang feel free to create a new PR!

**Comment by jon-chuang** [comment]
Ok! Merged the latest master here

**Pull request merged**
